### PR TITLE
os/filestore: handle EINTR returned by io_getevents()

### PR DESCRIPTION
**Pull request by liupan1111**
Signed-off-by: Pan Liu pan.liu@istuary.com

### CODE REVIEW

**Comment by tchaikov** [comment]
@liupan1111 please split this commit into three. one for removing "get_info_log_level()", one for handling EINTR return value from `io_getevents()`, and one for `ASSERT_EQ(0, fs.add_block_device(BlueFS::BDEV_DB, fn));`. i think we might want to hold the third one, and instead piggy back it with more significant changes.

**Comment by liupan1111** [comment]
@tchaikov  please split this commit into three. one for removing "get_info_log_level()", one for handling EINTR return value from io_getevents(), and one for ASSERT_EQ(0, fs.add_block_device(BlueFS::BDEV_DB, fn));. i think we might want to hold the third one, and instead piggy back it with more significant changes.

**Comment by liupan1111** [comment]
@tchaikov , thank you for suggestion!

Just not understand completely: I agree to split these changes into three. But how do I do this? Do you prefer I create three new PRs? Or a new PR with three commits?

**Comment by tchaikov** [comment]
a single PR would do. and you can reuse this PR to preserve the comment from reviewers. actually i'd suggest remove the third commit.

**Comment by liupan1111** [comment]
@tchaikov As you suggestion, I splitted it to 2 commits. Please take a look. Thanks.

**Comment by tchaikov** [comment]
could you prefix the title of your commit message with the subcomponent your are changing ? see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes

and remove the merge commit from your PR?

File: src/os/fs/FS.cc [hunk 1]
```diff
@@ -215,13 +215,28 @@ int FS::aio_queue_t::get_next_completed(int timeout_ms, aio_t **paio, int max)
     timeout_ms / 1000,
     (timeout_ms % 1000) * 1000 * 1000
   };
-  int r = io_getevents(ctx, 1, max, event, &t);
-  if (r <= 0) {
-    return r;
+
+  int r = 0;
+  while (1) {
+    r = io_getevents(ctx, 1, max, event, &t);
```
Comment (tchaikov) [inline src/os/fs/FS.cc]: could just put
```c
do {
   r = io_getevents(ctx, 1, max, event, &t);
} while (r == -EINTR);
```

Comment (liupan1111) [inline src/os/fs/FS.cc]: agree

**Comment by liupan1111** [comment]
@tchaikov, thank you for the change about the tile, I will study the link. I meet some issue with git, and will remove redudant merge commit ASAP.

**Comment by liupan1111** [comment]
@tchaikov , I've modified it as your suggestion, please help take a look. Thank you!

File: src/os/fs/FS.cc [hunk 2]
```diff
@@ -215,13 +215,20 @@ int FS::aio_queue_t::get_next_completed(int timeout_ms, aio_t **paio, int max)
...
+
+  int r = 0;
+  do {
+    r = io_getevents(ctx, 1, max, event, &t);
+  } while (r == -EINTR);
+  
   if (r <= 0) {
     return r;
   }
+
```
Comment (tchaikov) [inline src/os/fs/FS.cc]: please remove this line.

File: src/os/fs/FS.cc [hunk 3]
```diff
@@ -215,13 +215,20 @@ int FS::aio_queue_t::get_next_completed(int timeout_ms, aio_t **paio, int max)
...
   for (int i=0; i<r; ++i) {
     paio[i] = (aio_t *)event[i].obj;
     paio[i]->rval = event[i].res;
   }
+
```
Comment (tchaikov) [inline src/os/fs/FS.cc]: and this line.

Comment (liupan1111) [inline src/os/fs/FS.cc]: @tchaikov done. I will take care about this next time.

**Comment by liupan1111** [comment]
@tchaikov done. I will take care about this next time.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2016-11-16_11:34:58-rados-wip-kefu-testing---basic-mira/

**Pull request merged**